chart/values.yaml

```diff
@@ -319,15 +327,31 @@ meilisearch:
 postgresql:
   enabled: false
   auth:
-    # this does not read from a secret.  not sure how to map
-    # just manually make the same
-    username: "kyoo_back"
-    database: "{{ .Values.global.postgres.database }}"
+    # create a user
+    username: "{{ .Values.global.postgres.infra.user }}"
     existingSecret: "{{ .Values.global.postgres.infra.existingSecret }}"
     secretKeys:
+      # set the postgres user password to the same as our user
       adminPasswordKey: "{{ .Values.global.postgres.infra.passwordKey }}"
       userPasswordKey: "{{ .Values.global.postgres.infra.passwordKey }}"
   primary:
+    # create databases, schemas, and set search_path
+    initdb:
+      scripts:
+        # kyoo_back still requires public schema
+        # https://github.com/zoriya/Kyoo/issues/536
+        kyoo_back.sql: |
+          CREATE DATABASE {{ .Values.global.postgres.kyoo_back.database }} WITH OWNER {{ .Values.global.postgres.infra.user }}; 
+          \connect {{ .Values.global.postgres.kyoo_back.database }};
+          CREATE SCHEMA IF NOT EXISTS data AUTHORIZATION {{ .Values.global.postgres.infra.user }};
+        kyoo_transcoder.sql: |
+          CREATE DATABASE {{ .Values.global.postgres.kyoo_transcoder.database }} WITH OWNER {{ .Values.global.postgres.infra.user }};
+          \connect {{ .Values.global.postgres.kyoo_transcoder.database }};
+          REVOKE ALL ON SCHEMA public FROM PUBLIC;
+          CREATE SCHEMA IF NOT EXISTS data AUTHORIZATION {{ .Values.global.postgres.infra.user }};
```

Comment:
does thsi means you expect the schema's name should be the same as the user?

i guess i should add a config option for that?

Reply:
This just creates a schema named `data` and assigns the user `kyoo_all` as the owner.   Your application still creates a gocoder schema--but i did confirm it starts with this approach. 

Following best practices for database roles is easier done with a Kubernetes operator.  Zalando for instance will create database roles for OWNER, READWRITE, and READONLY.  Then assign users into those roles. If using multiple schemas, best practice is to also create OWNER, READWRITE, and READONLY per schema that can be assigned to users.  Having a database per microservice is just easier and avoids unneeded creation of additional roles.

In this implementation, we are creating one postgres instance using bitnami's helm chart.  The instance on setup will create two database based upon  `.Values.global.postgres.kyoo_back.database` and `.Values.global.postgres.kyoo_transcoder.database`.  kyoo_back requires the use of public schema as mentioned in ttps://github.com/zoriya/Kyoo/issues/536 so we need to leverage it.  kyoo_transcoder creates a schema named data to avoid ever using public then facing permission issues.

I get why you are having the application create the schema, because it's easier to work with in docker-compose.  My opinion is that those tasks should be part of the infrastructure setup. Application should just assume they already exist--this avoids the application needing to be aware of roles configuration and permission issues in general.

Reply:
the application creates them if they don't already exist, i don't think it conflicts with infra management of schemas, does it?

Reply:
I see where you’re coming from, and it makes sense that having the application create schemas as a fallback can provide extra resilience. However, my main concern is that it can blur the lines of responsibility between the infrastructure and the application. When the application manages schema creation, it introduces potential complexities with role management and permissions that might be better handled during setup. 

If the infrastructure is consistently responsible for creating the schemas, then application deployments become cleaner since they assume everything is pre-configured. But I agree that having the application create schemas as a safety net can still work as long as it doesn’t lead to conflicts or inconsistencies down the road.

Reply:
I don't think we have another good option to keep a simple docker-compose deployment. Do we?

Reply:
docker-compose based can be kept inline with this helm chart based initialization.  https://github.com/docker-library/docs/blob/master/postgres/README.md#initialization-scripts

have the postgres instance created without a default database, then use the following script to generate the databases.

```sql
CREATE DATABASE kyoo_backend WITH OWNER kyoo_all; 
\connect kyoo_backend;
CREATE SCHEMA IF NOT EXISTS data AUTHORIZATION kyoo_all;
CREATE DATABASE kyoo_transcoder WITH OWNER kyoo_all;
\connect kyoo_transcoder;
REVOKE ALL ON SCHEMA public FROM PUBLIC;
CREATE SCHEMA IF NOT EXISTS data AUTHORIZATION kyoo_all;
ALTER ROLE kyoo_all 
IN DATABASE kyoo_backend SET search_path TO "$user", public;
ALTER ROLE kyoo_all 
IN DATABASE kyoo_transcoder SET search_path TO "$user", data;
```

Using this, you can avoid creating a schema in the application and just rely upon what ever is in the search_path.

Reply:
The initialization scripts are only run once when the volume is created. This means it's impossible to add another schema when we add a new service. It doesn't seem like a good solution. I guess we could add an env var to disable schema handling in the app and let the infrastructure handle it tho.

Reply:
I hear you that it is a pain mamaging resources.  but I also think it is unreasonable for Kyoo to start specializing in configuring postgres.

A way to address this is to use advanced tooling like postgres operator which will handle the lifecycle of the databases and also roles/permissions.  

example snippet for configuring a database named mealie, a schema named data, and will also create database level user/roles for owner/read-write/read-only.

source: https://github.com/acelinkio/argocd-homelab/blob/main/manifest/postgres.yaml#L28

```yaml
    mealie:
      defaultUsers: true
      extensions: {}
      schemas:
        data:
          defaultRoles: false
          defaultUsers: false
```

As always, when using external subchart configuration in charts is best effort.  For example, when I updated the Meilisearch version from 0.7.0 to 0.8.0, that was a breaking change that required manual intervention.

Reply:
I understand the want to manually control postgres, but most docker-compose users don't care (or know) how to do this. This is unreasonable to ask them to manually create databases and schema. The default behavior for docker-compose needs to work without any manual efforts, and future updates needs to just work. This is also one of the reasons the default docker-compose is keeping postgres at version 15.

I'm open to any modifications that could make infra control easier for kube (or for the DX) but docker-compose users should not have to know or do anything.

Reply:
Honestly - I understand why you're going with multiple databases, but if simplicity like that is a goal, you're shooting yourself in the foot by taking that approach.

Reply:
This is typically what happens with microservice architecture. There is trade offs to each approach.  One way to avoid initialization complexity is to do a postgres instance per microservice.  The skilled power user can do consolidation if they wish.

Reply:
I feel like a default schema by application is a sane default, if someone wants to customise this they could override the schema via env var (or disable custom schema handling from the app)

something like

`POSTGRES_SCHEMA` that would default to `gocoder` but you could use anything.
If you specify `POSTGRES_SCHEMA=disabled` it would just use the default `search_path`.

What do you think?

Reply:
I merged the `POSTGRES_SCHEMA` handling!

Reply:
Updated helm chart to support specifying schema for transcoder.  Also added docs that encourage users to not rely upon sub-charts.  

Resolving conversation.  Please reopen if there are any questions or concerns.